src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/apply.h>
+#include <xrpld/app/tx/detail/Batch.h>
+#include <xrpld/ledger/Sandbox.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+TxConsequences
+Batch::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+
+    if (tx.getFlags() & tfBatchMask)
+    {
+        JLOG(ctx.j.warn()) << "Batch: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    AccountID const outerAccount = tx.getAccountID(sfAccount);
+
+    auto const& txns = tx.getFieldArray(sfRawTransactions);
+    STVector256 const hashes = tx.getFieldV256(sfTxIDs);
+    if (hashes.size() != txns.size())
+    {
+        JLOG(ctx.j.warn()) << "Batch: Hashes array size does not match txns.";
+        return temINVALID;
+    }
+
+    if (txns.empty())
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array empty.";
+        return temARRAY_EMPTY;
+    }
+
+    if (txns.size() > 8)
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array exceeds 12 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    for (int i = 0; i < txns.size(); ++i)
+    {
+        STTx const stx = STTx{STObject(txns[i])};
+        if (stx.getTransactionID() != hashes[i])
+        {
+            JLOG(ctx.j.warn()) << "Batch: Hashes array does not match txns.";
+            return temINVALID;
+        }
+
+        if (!stx.isFieldPresent(sfTransactionType))
+        {
+            JLOG(ctx.j.warn())
+                << "Batch: TransactionType missing in array entry.";
+            return temINVALID_BATCH;
+        }
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.warn()) << "Batch: batch cannot have inner batch txn.";
+            return temINVALID_BATCH;
+        }
+
+        auto const innerAccount = stx.getAccountID(sfAccount);
+        if (stx.getFieldU16(sfTransactionType) == ttACCOUNT_DELETE &&
+            innerAccount == outerAccount)
+        {
+            JLOG(ctx.j.warn())
+                << "Batch: inner txn cannot be account delete when inner and "
+                   "outer accounts are the same.";
+            return temINVALID_BATCH;
+        }
+
+        if (innerAccount != outerAccount)
+        {
+            if (!tx.isFieldPresent(sfBatchSigners))
+            {
+                JLOG(ctx.j.warn()) << "Batch: missing batch signers.";
+                return temBAD_SIGNER;
+            }
+
+            if (tx.getFieldArray(sfBatchSigners).end() ==
```

Comment:
There is no check on whether BatchSigners has too many signatures (repeats, other accounts, etc.)

Reply:
Or a max length check